[processing] Port refactor fields to new API

arnaud-morvan opened the pull request:
## Description

Port RefactorFields to new API.

In modeler, ParameterFieldsMapping values are store as QVariantList of QVariantMap : 
```
        <Option type="List" name="FIELDS_MAPPING">
          <Option type="Map">
            <Option type="int" value="2" name="source"/>
            <Option type="List" name="static_value">
              <Option type="Map">
                <Option type="QString" value="&quot;fid&quot;" name="expression"/>
                <Option type="int" value="0" name="length"/>
                <Option type="QString" value="fid" name="name"/>
                <Option type="int" value="0" name="precision"/>
                <Option type="int" value="10" name="type"/>
              </Option>
              <Option type="Map">
                <Option type="QString" value="&quot;Bname&quot;" name="expression"/>
                <Option type="int" value="4" name="length"/>
                <Option type="QString" value="Bname" name="name"/>
                <Option type="int" value="0" name="precision"/>
                <Option type="int" value="10" name="type"/>
              </Option>
              <Option type="Map">
                <Option type="QString" value="&quot;Bintval&quot;" name="expression"/>
                <Option type="int" value="0" name="length"/>
                <Option type="QString" value="Bintval" name="name"/>
                <Option type="int" value="0" name="precision"/>
                <Option type="int" value="2" name="type"/>
              </Option>
              <Option type="Map">
                <Option type="QString" value="&quot;Bfloatval&quot;" name="expression"/>
                <Option type="int" value="0" name="length"/>
                <Option type="QString" value="Bfloatval" name="name"/>
                <Option type="int" value="0" name="precision"/>
                <Option type="int" value="6" name="type"/>
              </Option>
            </Option>
          </Option>
        </Option>
```

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

arnaud-morvan committed:
[processing] Port refactor fields to new API

nyalldawson commented on python/plugins/processing/algs/qgis/FieldsMapper.py:
```diff
@@ -37,150 +36,89 @@
+                                        "",
+                                        field_def['length'],
+                                        field_def['precision']))
             expression = QgsExpression(field_def['expression'])
```
Any reason in particular you dropped the setting of geometry calculator/units from the expression? This will prevent expressions like `$area` from correctly using the project ellipsoid and unit settings.

arnaud-morvan replied:
I've restored the expression geometry calculator/units in the prepareAlgorithm now.

nyalldawson commented on python/plugins/processing/algs/qgis/FieldsMapper.py:
```diff
@@ -37,150 +36,89 @@
+
+    def processAlgorithm(self, parameters, context, feeback):
+        # create an expression context using thead safe processing context
+        self.expr_context = self.createExpressionContext(parameters, context)
```
I think after this you should prepare all the expressions from self.expressions against the context - it'll dramatically speed up their evaluation.

arnaud-morvan replied:
Done

nyalldawson commented on python/plugins/processing/algs/qgis/FieldsMapper.py:
```diff
@@ -37,150 +36,89 @@
+        attributes = []
+        for expression in self.expressions:
+            self.expr_context.setFeature(feature)
+            #expr_context.lastScope().setVariable("row_number", rownum)
```
you could make rownum a member of the alg here and increment after each feature is processed - that'll fix this up

nyalldawson reviewed (commented):
Fantastic!

arnaud-morvan committed:
[processing] Check all parameter's values in ModelerParametersDialog

Check that all parameter's values pass the validity check,
even if not returned as QgsProcessingModelChildParameterSource.

In case of list, tests that it is really a QgsProcessingModelChildParameterSource list,
and create a QgsProcessingModelChildParameterSource from the list if it is not the case
(useful for custom parameters that return lists as ParameterFieldsMapping).

arnaud-morvan commented:
@nyalldawson : I've treated your comments.
Note that I've set the ParameterFieldsMapping.type() to 'boolean' for the sip subclass, because without this QGIS crash (see https://github.com/arnaud-morvan/QGIS/blob/820e3422c81abc80ff4bab756cad1e868969250a/src/core/processing/qgsprocessingparameters.h#L188:L235). I do not really understand this, but it should be mandatory due to the virtual methods in QgsProcessingParameterDefinition.

arnaud-morvan commented:
Note that I'm currently working on Aggregate and I will soon push with small code improvement on FieldsMapper.

nyalldawson commented on python/plugins/processing/algs/qgis/FieldsMapper.py:
```diff
@@ -28,159 +27,141 @@
 
-            def setValue(self, value):
-                if value is None:
+            def checkValueIsAcceptable(self, value, context):
```
This is missing the default value for context - it should be `context=None` as the context is not always available or passed to this method (e.g. currently this crashes when running through the toolbox)

nyalldawson reviewed (commented):
This is working very nicely for me now.

One suggestion to improve the UI of the algorithm would be to replace the current `layerCombo` in the panel with a proper QgsMapLayerComboBox widget, instead of using the standard QComboBox and manually populating it. Then we'd get consistent behavior with the rest of QGIS and the nice layer type icons shown, tooltips, etc.

m-kuhn commented on src/core/processing/qgsprocessingparameters.h:
```diff
@@ -233,6 +233,8 @@ class CORE_EXPORT QgsProcessingParameterDefinition
     else if ( sipCpp->type() == QgsProcessingParameterBand::typeName() )
       sipType = sipType_QgsProcessingParameterBand;
+    else
+      sipType = NULL;
```
Can you update these to

`sipType = nullptr`

m-kuhn commented:
Is this ready for merge?
I'd love to see this algorithm ported.

arnaud-morvan commented:
@m-kuhn : Yes, for me this is ready for merge. I've now threated all comments, including nullPtr.
Once merge, I'll rebase Aggregate (#4210) on master.

m-kuhn commented:
Thanks a lot
